### ISSUE TITLE
Action error events behavior

### ISSUE THREAD

**Issue by marcelomorgado** [tasitlabs/TasitSDK#164]
Current behavior

```javascript
     txSubscription = simpleStorage.setValue("hello world");

      // AssertionError: expected [ 'error' ] to be empty 
      expect(txSubscription.subscribedEventNames()).to.be.empty; 

      txSubscription.on("confirmation", listener);
      txSubscription.off("confirmation");

      // AssertionError: expected [ 'error' ] to be empty
      expect(txSubscription.subscribedEventNames()).to.be.empty; 
```

Should these expectiations be correct?

Refs: https://github.com/tasitlabs/TasitSDK/pull/159#discussion_r250757612

**Comment by pcowgill** [comment]
One quick thought on issue naming. It should be easy for someone glancing at the project board to guess what an issue is about AND what the goal of it is in as few words as possible. When I read the name of this issue, I think "What about the behavior?"

**Comment by marcelomorgado** [comment]
OK, thanks.

**Comment by pcowgill** [comment]
In the current behavior, this is only the case if higher up in the code there's an error listener set for the contract, right? Or is there an error set for the `txSubscription` by default even without setting it up on the contract?

**Comment by pcowgill** [comment]
I think that's fine if it works that way. In a way it pushes users to remember to do something based on errors.

**Comment by marcelomorgado** [comment]
Another one:
If user subscribes twice:
```javascript
const s1 = contract.subscribe(); // instantiation
const s2 = contract.subscribe(); // reuse
```
Note: We decided keep `ContractEventsSubscription` as singleton.
In the case above, what is the expected behavior? one or two subscriptions?

And if the user does:
```javascript
const es = contract.subscribe();
const ts = contract.writeFunction();
contract.off("error");
contract.on("error", listener);
```
Should the contract re-subscribe to the `ContractEventsSubscription` and all `TransactionSubscription`s?

**Comment by pcowgill** [comment]
> Another one:
> If user subscribes twice:
> 
> ```js
> const s1 = contract.subscribe(); // instantiation
> const s2 = contract.subscribe(); // reuse
> ```
> 
> Note: We decided keep `ContractEventsSubscription` as singleton.
> In the case above, what is the expected behavior? one or two subscriptions?

First off, I think we may remove the `.subscribe` function.

Second, if we kept it, we could consider having the `.subscribe` function not return anything and just modify the class instance that calls it

> 
> And if the user does:
> 
> ```js
> const es = contract.subscribe();
> const ts = contract.writeFunction();
> contract.off("error");
> contract.on("error", listener);
> ```
> 
> Should the contract re-subscribe to the `ContractEventsSubscription` and all `TransactionSubscription`s?

I'll wait on answerign this until you respond to the point above.

**Comment by marcelomorgado** [comment]
Could we keep this thread waiting for now and them revisit this after https://github.com/tasitlabs/TasitSDK/issues/163 solved?

**Comment by pcowgill** [comment]
> Could we keep this thread waiting for now and them revisit this after #163 solved?

Yep!

**Comment by marcelomorgado** [comment]
@pcowgill commented:
> Does the subscription mind if the contract claims its only error listener? We should make a test for what happens if the developer then sets a different error listener explicitly on the subscription after first setting it on the contract

Refs: https://github.com/tasitlabs/TasitSDK/pull/159#discussion_r251036758

**Comment by marcelomorgado** [comment]
@pcowgill commented:
> Oh I get it, is this a hack to allow for two listeners, the catch-call one that the contract added and then the custom one?

@marcelomorgado  commented:
> Yes, it's necessary because we are allowing only one listener per event (error included).

Refs: https://github.com/tasitlabs/TasitSDK/pull/159#discussion_r251041163

**Comment by marcelomorgado** [comment]
Updating current behavior after [last changes](https://github.com/tasitlabs/TasitSDK/pull/166):

```javascript
contract.on("SomeEvent", listener);
console.log(contract.subscribedEventNames());
// [ "SomeEvent" ]

action = contract.setValue("hello world");
action.on("confirmation", listener);
console.log(action.subscribedEventNames());
// [ "error", "confirmation" ]
```

I'm not sure which path we should go.
Seems to me that the first behavior (no automatically error subscription) is more intuitive.

**Comment by pcowgill** [comment]
Hmm, yeah. I guess we could have that subscription maintained internally as not part of the list. And perhaps using a different event name than error, like "internal-error"

**Comment by marcelomorgado** [comment]
Which milestone this issue belongs?

**Comment by pcowgill** [comment]
> Which milestone this issue belongs?

@marcelomorgado I think `0.2.0`

**Comment by pcowgill** [comment]
@marcelomorgado Is this one still open?

**Comment by marcelomorgado** [comment]
I think so.
